### PR TITLE
Fixed module-based query

### PR DESCRIPTION
**Pull request by ssardina**
Fixes issue #9   to be able to use to post the query  in the correct module and not always in the default one contextModule (user0.

### CODE REVIEW

**Comment by ssardina** [comment]
@JanWielemaker , I have proposed this pull request to fix this bug, it is very simple but needs fixing I believe (unless I am missing something big)

**Comment by anionic** [comment]
I'm not a modules ninja but this looks plausible and I'm happy for it to be merged.

**Comment by ssardina** [comment]
Lines 316-329 does the reasoning on what is the module to be used, stored in var `module`.

But then it keeps using `contextModule` in [337](https://github.com/SWI-Prolog/packages-jpl/blob/0d14d8e4ce0e6c557e731d333e73ad0bc0b45855/src/java/org/jpl7/Query.java#L337). I think this is a small overlook, maybe the code above was added later to generalize to modules and they forgot to update the actual use! ;-)

**Comment by JanWielemaker** [comment]
Why would you want to set the context module to the module where you call the predicate?  Sure, using the fixed value `user` is most likely wrong.  If you pass `NULL` (at the C level), this will use the module of the innermost Prolog environment or `user` if this doesn't exist.

Can your use case in more detail?

**Comment by ssardina** [comment]
Yes, of course Jan. 

I am using SWI as the Knowledge Base for a multi-agent system done in Java-based [SARL](http://www.sarl.io/).

To allow each agent to have its own "beliefs"/knowledge-base, I use SWI modules: each agent will assert and query its module (module named with the id of the agent). That is why I need to qualify every query done: if it is done from agent X, it must be done to module X.

Does it make sense?

**Comment by JanWielemaker** [comment]
> Does it make sense?

Possibly.  Let me rephrase to see I got it.  You have a module `A1` (representing the knowledge of
agent 1.  This module imports some rule `R` that is a meta-predicate.   You call this `R` from Java, where `R` in turn must call facts and rules in `A1`.  

Is that what happens?  Surely that would make sense.  What about the following:

  - The default context module passed for PL_open_query() is NULL.
  - If the query is of the form `m:g`, the context module defaults to `m`
  - Ideally, I guess, it should be possible to set an explicit context module.  I don't know how that is typically done in Java.

Agree?  If so, could you turn that into a pull request?   My Java is really poor ...

**Comment by ssardina** [comment]
Exactly. And that is what we do, we issue queries of the form `m:g` where `m` is the id of the agent.

Now in line (316](https://github.com/SWI-Prolog/packages-jpl/blob/0d14d8e4ce0e6c557e731d333e73ad0bc0b45855/src/java/org/jpl7/Query.java#L316), the code is already there to extract `m`:

```
if (goal_.hasFunctor(":", 2)) {
			if (goal_.arg(1).isAtom()) {
				module = goal_.arg(1).name();
			}
```

otherwise, if there is no `:`, then we default to `module = contextModule;`

However, down below in line [337](https://github.com/SWI-Prolog/packages-jpl/blob/0d14d8e4ce0e6c557e731d333e73ad0bc0b45855/src/java/org/jpl7/Query.java#L337) a query is opened in `contextModule` isntead of `module`.

That is the fix of my request, which I believe is already a pull request Jan, isn't it?

**Comment by JanWielemaker** [comment]
The `m:g` case is fine.  The default however is `user`, while it should be `NULL` (at least, I think it should be) and I guess it makes sense if the user can change `contextModuie` before opening the query.  I think that is not the case (or can anyone just assign this?)  I don't know what needs to be done to get `NULL` as default to the C layer.

If you think that is beyond you, I can merge this pull request as it is surely a step in the right direction.

**Comment by ssardina** [comment]
ohh I see what you are referring now. Do you mean these lines are not that right?:

```
	protected final String hostModule = "user"; // until revised constructors
	protected final String contextModule = "user"; // until revised constructors

```

OK I see. yeah, I don't know enough yet about how to do NULL well and why `user` is there... What wold be the effect of passing NULL to `PL_open_query(..)`.

Nonetheless seems this is a related but orthogonal issue to line 337, as you hinted. even if we change that `user`, like 337 has to change to `module` to support when the query is done against a particular module right?

**Comment by ssardina** [comment]
Thanks. 

OK, good point. I will investigate this NULL vs user issue more and let you know. I need to understand first what is a "Prolog environment" from Java amount to. I have always assumed JPL will create one "engine" and that the default from the Java side would be `user`. But your text seems to suggest a Java Prolog environment can already be in a given module. mmm I need to check that Prolog class. Will do.

Thanks for the merge again, all good with that bit.

**Comment by JanWielemaker** [comment]
The environment comes into play when Prolog calls Java and Java calls back to Prolog.

**Comment by ssardina** [comment]
I am curious to know: is this small "fix" part of the latest SWI? If so, for all platforms including Windows (I use Linux but some students may use Windows so I want to know..)? 

I couldn't find any reference to this in the [Changelog](http://www.swi-prolog.org/ChangeLog?branch=stable) so want to double check this (not sure if Changelog also includes the extra packages)

**Comment by ssardina** [comment]
@JanWielemaker, I wanted to know if this fix in JPL has been incorporated into the official stable versions of SWI, and seems not from the source code of stable SWI. (I am currently using a fork of this repo myself with the fix, but I would much rather use the official version of SWI. )

My understanding is that the fix will appear in the 7.7.x versions, but SWI is still in the 7.6.x. Is this understanding correct?
